sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CollapseRepartitionSuite.scala

```diff
@@ -43,15 +55,44 @@ class CollapseRepartitionSuite extends PlanTest {
     comparePlans(optimized, correctAnswer)
   }
 
+  test("collapse one coalesce and one repartition into one") {
+    val query1 = testRelation
+      .coalesce(20)
+      .repartition(5)
+
+    val optimized1 = Optimize.execute(query1.analyze)
+    val correctAnswer1 = testRelation.repartition(5).analyze
+
+    comparePlans(optimized1, correctAnswer1)
+
+    val query2 = testRelation
+      .repartition(5)
+      .coalesce(20)
+
+    val optimized2 = Optimize.execute(query2.analyze)
+    val correctAnswer2 = testRelation.repartition(5).coalesce(20).analyze
```

Comment:
For improving this rule, we need to clean up the resolution of RepartitionByExpression at first.